File: main/scala/dynamicprogramming/DiceProblem.scala
package dynamicprogramming

/*
Given a number 'n', n>=0, find out the number of ways the number 'n' can be reached
by rolling a 6-faced dice
 */

object DiceProblem extends App {

  def noOfWays(n : Int) : BigInt = n match {
    case 0 => 0
    case a if a <= 6 => 1 + (1 to a).map(i => noOfWays(a - i)).sum
    case _ =>
      val array = new Array[BigInt](n+1)
      array(0) = 0
      (1 to 6).foreach(i => array(i) = noOfWays(i))
      (7 to n).foreach( i =>
        array(i) = array(i-6) + array(i-5) + array(i-4) + array(i-3) + array(i-2) + array(i-1))
      array(n)
  }


}

File: haskell-way/src/DiceProblem.hs
{-# LANGUAGE MultiWayIf #-}

module DiceProblem where

import Data.Function.Memoize (memoFix)

{-
  Given a number 'n', n>=0, find out the number of ways the number 'n' can be reached
by rolling a 6-faced dice
-}

{-
  This problem could be solved in Haskell using a linked list, similar to the eaxmple in Scala. It would be more idiomatic to 
-}

{- TODO: Original function, implemented with linked lists as in scala-}
noOfWays :: Int -> Integer
noOfWays = undefined

noOfWays' :: Int -> Integer
noOfWays' = memoFix $ \memo n ->
  if | n <= 0    -> 0
     | n <= 6    -> 1 + sum [ memo (n -i) | i <- [1..n] ]
     | otherwise -> sum [ memo (n - i) | i <- [1..6] ]